hack/test-go.sh

```diff
@@ -1,125 +1,231 @@
-  fi
+# Internalize envionment variables we consume and default if they're not set
+print_packages="${PRINT_PACKAGES:-}"
+test_kube="${TEST_KUBE:-}"
+test_timeout="${TIMEOUT:-60s}"
+detect_races="${DETECT_RACES:-true}"
+coverage_output_dir="${COVERAGE_OUTPUT_DIR:-}"
+coverage_spec="${COVERAGE_SPEC:--cover -covermode atomic}"
+gotest_flags="${GOTEST_FLAGS:-}"
+junit_report="${JUNIT_REPORT:-}"
+
+if [[ -n "${junit_report}" && -n "${coverage_output_dir}" ]]; then
+    echo "$0 cannot create jUnit XML reports and coverage reports at the same time."
+    exit 1
 fi
 
-if [ -z ${KUBE_RACE+x} ]; then
-  KUBE_RACE=""
+# determine if user wanted verbosity
+verbose=
+if echo "${gotest_flags}" | grep -Eq "\-v( |$)"; then
+    verbose=true
 fi
 
-KUBE_TIMEOUT=${KUBE_TIMEOUT:-'60s'}
-
-if [ "${1-}" != "" ]; then
-  suffix=/...
-  if [[ "${1}" == *"${suffix}" ]]; then
-    # if the package name ends with /... they are intending a recursive test
-    test_packages=$(find_test_dirs "${1%$suffix}")
-  else
-    test_packages="$OS_GO_PACKAGE/$1"
-  fi
-elif [ -n "$TEST_KUBE" ]; then
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs; find_upstream_test_dirs`
-else
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs`
+# Build arguments for 'go test'
+if [[ -z "${verbose}" && -n "${junit_report}" ]]; then
+    # verbosity can be explicitly set or set implicitly by asking for the jUnit XML report, 
+    # so we only want to add the flag if it hasn't been added by a user already and is being
+    # implicitly set by jUnit report generation
+    gotest_flags="${gotest_flags} -v"
 fi
 
-if [ -n "$PRINT_PACKAGES" ]; then
-  for package in $test_packages
-  do
-    echo $package
-  done
+if [[ -n "${detect_races}" ]]; then
+    gotest_flags="${gotest_flags} -race"
+fi
 
-  exit 0
+# check to see if user has not disabled coverage mode
+if [[ -n "${coverage_spec}" ]]; then
+    # if we have a coverage spec set, we add it. '-race' assumes '-cover -covermode atomic'
+    # but specifying both at the same time does not lead to an error so we can add both specs
+    gotest_flags="${gotest_flags} ${coverage_spec}"
 fi
 
-export OPENSHIFT_ON_PANIC=crash
+# check to see if user has not disabled test timeouts
+if [[ -n "${test_timeout}" ]]; then
+    gotest_flags="${gotest_flags} -timeout ${test_timeout}"
+fi
 
-if [[ -n "${KUBE_COVER}" && -n "${OUTPUT_COVERAGE}" ]]; then
-  # Iterate over packages to run coverage
-  test_packages=( $test_packages )
-  for test_package in "${test_packages[@]}"
-  do
-    mkdir -p "$OUTPUT_COVERAGE/$test_package"
-    KUBE_COVER_PROFILE="-coverprofile=$OUTPUT_COVERAGE/$test_package/profile.out"
+# list_test_packages_under lists all packages containing Golang test files that we want to run as unit tests
+# under the given base dir in the OpenShift Origin tree
+function list_test_packages_under() {
+    local basedir=$1
+
+    find ${basedir} -not \( \
+        \( \
+              -wholename 'Godeps' \
```

Comment:
find(1) says that `-wholename` is a less portable synonym for `-path`, so maybe it would be better to use the latter.  (Note that `find` is not a builtin, so there may be a higher chance that we'll get an unusual `find`.)